### PR TITLE
Handle AdoptFail state in Ironic Adopt 

### LINKED ISSUE #697: Retry when node in "Adopt Failed" state
Currently, we are not retrying when ironic fails to adopt and puts the node in 'adopt failed' state:  https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L1001

As [checked](https://docs.openstack.org/ironic/latest/_images/states.svg), we could try to retry to get a node out of "adopt failed" by:

-  sending manage command(via API) ==> manageable
- sending adopt command (via API) ==> adopting(retry adoption)

### PR DESCRIPTION
**Pull request by furkatgofurov7**
Fixes #697

### CODE REVIEW

File: pkg/provisioner/ironic/ironic.go [hunk 1]
```diff
@@ -1025,8 +1025,20 @@ func (p *ironicProvisioner) Adopt() (result provisioner.Result, err error) {
 		result.RequeueAfter = provisionRequeueDelay
 		result.Dirty = true
 	case nodes.AdoptFail:
-		result.ErrorMessage = fmt.Sprintf("Host adoption failed: %s",
-			ironicNode.LastError)
+		// The node is in AdoptFail, the way to get node
+		// out of this state is by target manage to go
+		// to manageable. Once the node is in manageable
+		// it will move to target adopt.
+		result, err = p.changeNodeProvisionState(
```
Comment (maelk) [inline pkg/provisioner/ironic/ironic.go:1032]: I'm concerned that if this fails we will requeue for the backoff delay, and when coming back here, we will still be in AdoptFail, so we will go to manageable, and requeue again for the backoff time, making it in effect two successive backoffs. In the function that calls this function, can we requeue immediately if err is not nil ?

File: pkg/provisioner/ironic/adopt_test.go [hunk 2]
```diff
@@ -73,6 +73,7 @@ func TestAdopt(t *testing.T) {
 
 			expectedDirty:        false,
 			expectedRequestAfter: 0,
+			expectedError:        true,
 		},
 	}
```
Comment (jan-est) [inline pkg/provisioner/ironic/adopt_test.go:79]: Do we still need a test case for retry and to make sure that Node will move from AdoptFail state back to manageable?

File: pkg/provisioner/ironic/ironic.go (hunk 1 above)
Comment (furkatgofurov7) [reply pkg/provisioner/ironic/ironic.go:1032]: So, do you mean, we could call requeue [here](https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/baremetalhost_controller.go#L691)?

**Comment by zaneb** [comment]
What do you think of https://github.com/zaneb/baremetal-operator/commit/36fd3fb5a533ee9b7d6b5d92a3ab2c7d7ff01d5c?

**Comment by maelk** [comment]
@zaneb your commit looks good. @furkatgofurov7 would you be ok with it ?

**Comment by furkatgofurov7** [comment]
@maelk sure, I am okay with that. @zaneb will you open a PR then with your commit, I will close this one? Thanks!

**Comment by zaneb** [comment]
Yep, it's part of a series that I still need to test, but will open a PR as soon as that's ready.

**Comment by furkatgofurov7** [comment]
> Yep, it's part of a series that I still need to test, but will open a PR as soon as that's ready.

@zaneb thanks! Closing this, as it will be covered by another patch. 

/close

**Pull request closed without merging**